File: validation.js
import {create_card_obj} from "./utils.js";
import {TARGETS} from "./constants.js";

function is_valid_pile_drop(pile_id, card_str, state) {
    const pile = state[pile_id]
    const pile_card = create_card_obj(pile.at(-1))
    const card = create_card_obj(card_str)

    if (card.rank === "A") return false;
    if (pile.length === 0) return true;
    if (card.suit_value === pile_card.suit_value) return false;
    if (card.rank_value + 1 !== pile_card.rank_value) return false;

    return true;
}

function is_valid_stack_drop(stack_id, card_str, state) {
    const stack = state[stack_id]
    const stack_card = create_card_obj(stack.at(-1))
    const card = create_card_obj(card_str)

    if (stack.length === 0) return card.rank_value === 1;
    if (stack_card.rank_value + 1 === card.rank_value
        && stack_card.suit === card.suit) return true;

    return false;
}

function is_valid_opponent_drop(opponent_id, card_str, state) {
    const opponent = state[opponent_id].at(-1);
    const opponent_card = create_card_obj(opponent);
    const card = create_card_obj(card_str);

    if (!opponent_card) return false;
    if ((opponent_card.rank_value + 1 === card.rank_value
        || opponent_card.rank_value - 1 === card.rank_value
        ) && opponent_card.suit === card.suit)
        return true;

    return false;
}

function is_valid_move(target_id, card, state, target, src) {
    if (state.turn_player_id !== state.player_id) return false;
    if (state.is_card_drawn && !src.startsWith("player_pile")) return false;
    if (TARGETS.pile === target) return is_valid_pile_drop(target_id, card, state);
    if (TARGETS.stack === target) return is_valid_stack_drop(target_id, card, state);
    if (TARGETS.opponent_pile === target || TARGETS.opponent_reserve === target) {
        return is_valid_opponent_drop(target_id, card, state);
    }
    console.error("Err: Unknown drop target {is_valid_move}");
    return false;
}

function is_player_turn(state) {
    return state.turn_player_id === state.player_id;
}

function is_valid_turn_end(state) {
    if ((state.player_cards_len[1] <= 1 && state.player_cards_len[2] === 0) || state.is_stop) return true;
    return state.is_card_drawn;
}

export  {
    is_valid_move,
    is_player_turn,
    is_valid_turn_end
}

File: Server/Validation.cs
namespace Server;
using static Server.Constants;

public static class Validation
{
    public static bool IsValidMove(MoveMessage incoming, ref GameState state, bool isPlayer)
    {
        var src = State.GetList(incoming.Src, ref state);
        var dst = State.GetList(incoming.Target, ref state);
        if (src is null || dst is null ) { Console.WriteLine("Could not find src/dst list"); return false; }
        if (src.Count == 0) return false;

        if (state.IsCardDrawn && (
                (isPlayer && !incoming.Src.Equals("player_pile")) ||
                (!isPlayer && !incoming.Src.Equals("opponent_pile"))
                )) return false;
        if (incoming.Target.StartsWith("pile")) return IsValidPileDrop(src, dst);
        if (incoming.Target.StartsWith("stack")) return IsValidStackDrop(src, dst);
        if (incoming.Target.StartsWith("opponent") || incoming.Target.StartsWith("player")) 
            return IsValidPlayerDrop(src, dst);
        
        return false;
    }
    
    static bool IsValidPileDrop(List<Card> src, List<Card> dst)
    {
        var srcCard = src.Last();
        var dstCard = dst.LastOrDefault();
        
        if (srcCard.RankValue == RANK_VALUE.A) return false;
        if (dst.Count == 0) return true;
        if (srcCard.SuitValue == dstCard.SuitValue) return false;
        if (srcCard.RankValue >= dstCard.RankValue) return false;
        
        return true;
    }
    
    static bool IsValidStackDrop(List<Card> src, List<Card> dst)
    {
        var srcCard = src.Last();
        var dstCard = dst.LastOrDefault();
        
        if (dst.Count == 0) return (int) src.Last().RankValue == 1;
        if (srcCard.RankValue == dstCard.RankValue + 1 && srcCard.Suit == dstCard.Suit) return true;
        
        return false;
    }
    
    static bool IsValidPlayerDrop(List<Card> src, List<Card> dst)
    {
        var srcCard = src.Last();
        var dstCard = dst.LastOrDefault();

        if (dst.Count == 0) return false;
        if ((srcCard.RankValue + 1 == dstCard.RankValue || srcCard.RankValue -1 == dstCard.RankValue)
            && srcCard.Suit == dstCard.Suit) 
            return true;
        
        return false;
    }
    
    public static bool IsValidTurnEnd(Connection conn, ref GameState state)
    {
        var pileLength = conn.TurnId == 0 ? state.PlayerPile.Count : state.OpponentPile.Count;
        var deckLength = conn.TurnId == 0 ? state.PlayerDeck.Count : state.OpponentDeck.Count;
        
        if (pileLength <= 1 && deckLength == 0) return true;
        return state.IsCardDrawn;
    }

    public static bool IsValidDrawCard(ref GameState state) => state.IsCardDrawn;
    public static bool IsPlayerTurn(Connection conn, ref GameState state) => state.TurnPlayerId == conn.TurnId;
}

